#682 Improved documentation about coordinates and units.

rmjarvis opened the pull request:
This PR addresses issues #682 and #657, both related to documentation.

For #682, I added documentation about what our convention for x,y is in terms of the rows and columns in the numpy array. I also added some concrete examples that will hopefully make clear what some of the image methods that shift around the coordinates do.  @esheldon Please see if the documentation update here addresses your concerns.

For #657, I tried to improve the documentation about the units used in Airy, OpticalPSF and Kolmogorov.  In particular with respect to the units of the `lam_over_diam` and `lam_over_r0` parameters.  I also added a calculation of lam / r0, giving those values directly for Kolmogorov, analogous to the lam / diam calculation that Rachel added to Airy and OpticalPSF.  @dkirkby Please see if this adequately addresses the concerns you raised about the units of these parameters.

In addition, I added some more documentation about the fact that GSObjects are always in sky coordinates.  In particular that `GSObject.shift` is necessarily in terms of sky coordinates, since the GSObject doesn't know about the pixel scale yet, but that `drawImage(offset)` is in terms of pixels.  This is a question that has come up a number of times, so it seems like our documentation on this had been insufficient.

dkirkby commented:
This looks good to me, thanks.

rmandelb commented on galsim/base.py:
```diff
@@ -58,13 +58,21 @@ class GSObject(object):
+    For instance, if you eventually draw onto an image that has a pixel scale of 0.2 arcsec/pixel,
+    then the normal thing to do would be to define your surface brightness profiles in terms of
+    arcsec and then draw with `pixel_scale=0.2`.  However, nowhere in GalSim do we enforce that sky
+    coordinates need to be in arcsec.  If you wanted, you could instead define the sizes of all
```
Don't we have a few places where we enforce units, for example in the use of the COSMOS catalog? (noise pad size, target pixel size, half-light radius cuts).  Also NFW halo, and ChromaticAtmosphere.

rmjarvis replied:
- COSMOSCatalog: Yes, we should at least add some documentation that the objects built by COSMOSCatalog will have their linear dimensions in arcsec.  This seems to be poorly documented at the moment.
- NFW halo: I think the documentation on this is pretty good already that the positions should be given in arcsec.  We have a hook for units in a bunch of places to extend it to allow other units, although we say that only arcsec are implemented so far.  So I think this is ok, but there is room for converting them to use a similar kind of conversion calculation as we do in these classes.
- ChromaticAtmosphere: It would be easy to convert this to take a scale_unit parameter like these classes.  The arcsec is explicitly implemented [here](https://github.com/GalSim-developers/GalSim/blob/master/galsim/chromatic.py#L1113), so it would be very easy to divide by scale_unit instead.  I'll make that change.

rmjarvis replied:
COSMOSCatalog: Done d12e907
ChromaticAtmosphere: Done 0ab1f68

rmjarvis replied:
And I changed the language here to be less dogmatic about nowhere enforcing arcsec.  21796d1

rmandelb replied:
Thanks - the change in language here was what I had in mind, but the other changes are good!  However, for the ChromaticAtmosphere change, since you added a bit of functionality, perhaps it would be good to use that functionality in a unit test?

rmandelb commented on galsim/base.py:
```diff
@@ -1732,15 +1759,14 @@ class Airy(GSObject):
-
-        >>> airy = galsim.Airy(lam=lam, diam=diam)
+    To make this process a bit simpler, we recommend instead providing the wavelength and diameter
+    separately using the `lam` and `diam` parameters.  GalSim will then convert this to any of
```
mention these are required to be in units of nm and m, respectively?

rmjarvis replied:
Done: dfa4cd5

rmandelb commented:
Other than my nitpicky comments above, this looks good to me.

jmeyers314 commented on galsim/base.py:
```diff
@@ -1887,30 +1911,56 @@ class Kolmogorov(GSObject):
+    @param lam              Lambda (wavelength) in units of nanometers.  Must be supplied with
+                            `r0`, and in this case, image scales (`scale`) should be specified in
+                            units of `scale_unit`.
+    @param r0               The Fried parameter in units of meters.  Must be supplied with `lam`,
```
When constructing the realistic atmospheric PSF code, I found it convenient to have a `r0_500` argument, the Fried parameter at 500nm.  This made it conceptually easier to investigate wavelength dependence of the PSF, since I could look through the same atmosphere at different wavelengths just by changing `lam` instead of needing to change both `lam` and `r0`.  I think having both options would be reasonable here (and in the real atmosphere code; need to do that).  I'll stop short of recommending a `lam_over_r0_500` parameter, though.  That seems too complicated.

rmjarvis replied:
Good idea.  Done.  b33f8d6

jmeyers314 commented:
Looks go to me, just the one minor comment above.

rmandelb commented:
Any reason not to merge this one?

rmjarvis commented:
I guess I was waiting for @esheldon to take a look.  He's been pretty sick for the past month though.  Erin, did you want to see if this addresses your concerns about col,row vs x,y?  I don't mind waiting if you don't have time yet, but want to get to it in a week or so.

esheldon commented:
What I've read is helpful.  I don't know how to check if the (x,y)->(col,row) convention is documented comprehensively throughout the code, but these examples are good.

rmjarvis commented:
OK.  Thanks.  If down the line you notice a doc string that still seems confusing to you, let us know.
